### PR TITLE
Added concat BF16/FP32 BWD OneDNN kernel

### PR DESCRIPTION
**Pull request by jakpiase**
### PR types
New features

### PR changes
OPs

### Describe
Added concat BF16/FP32 BWD OneDNN kernel. I have also done some refactoring with concat tests

### CODE REVIEW

File: paddle/fluid/operators/concat_op.cc [hunk 1]
```diff
@@ -169,9 +169,18 @@ class ConcatOpGrad : public framework::OperatorWithKernel {
...
       const framework::ExecutionContext &ctx) const override {
-    return framework::OpKernelType(OperatorWithKernel::IndicateVarDataType(
-                                       ctx, framework::GradVarName("Out")),
-                                   ctx.GetPlace());
+    auto input_data_type = OperatorWithKernel::IndicateVarDataType(
+        ctx, framework::GradVarName("Out"));
+
+#ifdef PADDLE_WITH_MKLDNN
+    if (ctx.HasAttr("use_mkldnn"))  // fix for weird reverse op behavior
+      if (this->CanMKLDNNBeUsed(ctx, input_data_type)) {
```
Comment (lidanqing-vv) [inline paddle/fluid/operators/concat_op.cc]: Is this comment here valid, could you clarify more, Thank you very much.
Also there are two "if" you may do "if ( 1 && 2 )"?

File: python/paddle/fluid/tests/unittests/mkldnn/test_concat_mkldnn_op.py [hunk 2]
```diff
@@ -15,78 +15,137 @@
...
+class TestConcatAxis3OneDNNOp(TestConcatAxis0OneDNNOp):
+    def init_axis(self):
+        self.axis = 3
+
+    def init_shape(self):
+        self.x0_shape = [5, 3, 5, 5]
+        self.x1_shape = [5, 3, 5, 6]
+        self.x2_shape = [5, 3, 5, 7]
+
+
```
Comment (lidanqing-vv) [inline python/paddle/fluid/tests/unittests/mkldnn/test_concat_mkldnn_op.py:101]: You may add
```
# --------------------test concat bf16 in with axis 0,1,2, 3--------------------
```
at this location. And put `#----- concat fp32 ....----` comment on top of `TestConcatAxis0OneDNNOp` .

**Comment by lidanqing-vv** [comment]
@lidanqing-intel @wzzju Hi, this PR need approval. Since now concat grad fp32/bf16 op is added, some UT could be combined together.

**Review by lidanqing-vv — DISMISSED** [review]
LGTM

**Review by jczaja — DISMISSED** [review]
LGTM

**Review by wzzju — DISMISSED** [review]
LGTM

**Review by jczaja — APPROVED** [review]

**Pull request merged**